File: Compression.java
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.zip.DataFormatException;
import java.util.zip.Deflater;
import java.util.zip.GZIPOutputStream;
import java.util.zip.Inflater;

public class Compression {
    public static void main(String[] args) throws DataFormatException, FileNotFoundException, IOException {
        String inputString = "blahblahblahblahblah";
        System.out.println(inputString);
        byte[] inputBytes = inputString.getBytes(StandardCharsets.UTF_8);
        System.out.println("Uncompressed length: " + inputBytes.length);

        // Compress the bytes directly to a .gz file
        GZIPOutputStream gzos = new GZIPOutputStream(new FileOutputStream("compressed.gz"));
        gzos.write(inputBytes);
        gzos.close();

        // Compress the bytes
        byte[] outputBytes = new byte[100];
        Deflater compressor = new Deflater();
        compressor.setInput(inputBytes);
        compressor.finish();
        int compressedLength = compressor.deflate(outputBytes);
        compressor.end();
        System.out.println("Compressed length: " + compressedLength);
        // for (int i = 0; i < compressedLength; i++) {
        //     String st = String.format("%02x", outputBytes[i]);
        //     System.out.print(st);
        // }
        // System.out.println();

        // Decompress the bytes
        Inflater decompressor = new Inflater();
        decompressor.setInput(outputBytes, 0, compressedLength);
        byte[] result = new byte[100];
        int uncompressedLength = decompressor.inflate(result);
        decompressor.end();

        // Decode the bytes into a String
        System.out.println(new String(result, 0, uncompressedLength, StandardCharsets.UTF_8));
    }
}

File: compression.c
#include <stdio.h>
#include <string.h>
#include <zlib.h>

int main(void)
{
    gzFile file;
    z_stream defstrm, infstrm;
    unsigned char outputBytes[100];
    char result[100];

    const char *inputString = "blahblahblahblahblah";
    printf("%s\n", inputString);
    printf("Uncompressed length: %zu\n", strlen(inputString));

    // Compress the bytes directly to a .gz file
    file = gzopen("compressed.gz", "wb");
    gzwrite(file, inputString, strlen(inputString)); // same as gzputs(file, inputString);
    gzclose(file);

    // Compress the bytes
    defstrm.zalloc = Z_NULL;
    defstrm.zfree = Z_NULL;
    defstrm.opaque = Z_NULL;
    defstrm.avail_in = strlen(inputString);
    defstrm.next_in = (Bytef *) inputString;
    defstrm.avail_out = sizeof(outputBytes);
    defstrm.next_out = outputBytes;
    deflateInit(&defstrm, Z_BEST_COMPRESSION);
    deflate(&defstrm, Z_FINISH);
    deflateEnd(&defstrm);
    // printf("Compressed length: %lu\n", defstrm.total_out);
    // for (unsigned long i = 0; i < defstrm.total_out; i++) {
    //     printf("%02x", outputBytes[i]);
    // }
    // printf("\n");

    // Decompress the bytes
    infstrm.zalloc = Z_NULL;
    infstrm.zfree = Z_NULL;
    infstrm.opaque = Z_NULL;
    infstrm.avail_in = defstrm.total_out;
    infstrm.next_in = outputBytes;
    infstrm.avail_out = sizeof(result);
    infstrm.next_out = (Bytef *) result;
    inflateInit(&infstrm);
    inflate(&infstrm, Z_NO_FLUSH);
    inflateEnd(&infstrm);

    // Decode the bytes to a string
    result[infstrm.total_out] = '\0';
    printf("%s\n", result);
}
